### PR TITLE
feat: use `SortedSet` in model to improve reproducibility

### PR DESCRIPTION
**Pull request by RodneyRichardson**
Added `__lt__()` to all model classes used in SortedSet, with tests
Explicitly declared Enums as (str, Enum) to allow sorting
Added dependency to sortedcollections package
Added ComparableTuple to make sorting compound objects easier

Signed-off-by: Rodney Richardson <rodney.richardson@cambridgeconsultants.com>

### CODE REVIEW

**Comment by jkowalleck** [comment]
@RodneyRichardson  besides those lift dindings, 
please fix the findings of mypy and other tools.

* see https://github.com/CycloneDX/cyclonedx-python-lib/blob/main/CONTRIBUTING.md
* see https://github.com/CycloneDX/cyclonedx-python-lib/runs/6626778353?check_suite_focus=true
* see https://github.com/CycloneDX/cyclonedx-python-lib/runs/6626778287?check_suite_focus=true
* see https://github.com/CycloneDX/cyclonedx-python-lib/runs/6626778036?check_suite_focus=true

**Comment by RodneyRichardson** [comment]
> @RodneyRichardson besides those lift findings, please fix the findings of mypy and other tools.
> 
> * see https://github.com/CycloneDX/cyclonedx-python-lib/blob/main/CONTRIBUTING.md
> * see https://github.com/CycloneDX/cyclonedx-python-lib/runs/6626778353?check_suite_focus=true
> * see https://github.com/CycloneDX/cyclonedx-python-lib/runs/6626778287?check_suite_focus=true
> * see https://github.com/CycloneDX/cyclonedx-python-lib/runs/6626778036?check_suite_focus=true

I've fixed all those, I think. I have an outstanding mypy error I don't know how to resolve:
https://github.com/CycloneDX/cyclonedx-python-lib/runs/6628435343?check_suite_focus=true
```
cyclonedx/model/__init__.py:62: error: Missing type parameters for generic type
"tuple"  [type-arg]
    class ComparableTuple(tuple):
```
I've tried replacing `tuple` with `tuple[T,...]`, but I get a false positive about unreachable code. Any thoughts?

PS: Do the lift findings automatically resolve when I commit? Or do I need to address them individually above?

**Comment by RodneyRichardson** [comment]
@jkowalleck Ah - it also seems to be failing on python 3.8 and earlier! Any thgouhts?

```
File "/home/runner/work/cyclonedx-python-lib/cyclonedx-python-lib/cyclonedx/model/__init__.py", line 437, in <module>
    class ExternalReference:
  File "/home/runner/work/cyclonedx-python-lib/cyclonedx-python-lib/cyclonedx/model/__init__.py", line 499, in ExternalReference
    def hashes(self) -> SortedSet[HashType]:
TypeError: 'ABCMeta' object is not subscriptable
```

File: tests/test_model_component.py [hunk 1]
```diff
@@ -366,6 +424,29 @@ def test_not_same_1(self) -> None:
         self.assertNotEqual(id(p1), id(p2))
         self.assertFalse(p1 == p2)
 
+    def test_sort(self) -> None:
+        diff_a = Diff(text='a')
+        diff_b = Diff(text='b')
```
[the remark this answers is not in the document]
Comment (RodneyRichardson) [reply tests/test_model_component.py:429]: I need to fix this.

**Comment by jkowalleck** [comment]
@RodneyRichardson 


re: 
> cyclonedx/model/__init__.py:62: error: Missing type parameters for generic type "tuple"  [type-arg]
>    class ComparableTuple(tuple):

the typing.Tuple might help.
```py
from typing import Tuple

class Foo(Tuple[TypeOfItem1, and so on ]): 
  ...
```

----

regarding `SortedSet` and typing i cannot help.

File: cyclonedx/model/__init__.py [hunk 2]
```diff
@@ -57,7 +59,41 @@ def sha1sum(filename: str) -> str:
     return h.hexdigest()
 
 
-class DataFlow(Enum):
+class ComparableTuple(tuple):
+    """
+    Allows comparison of tuples, allowing for None values.
+    """
+
+    def __lt__(self, other: object) -> bool:
```
[the remark this answers is not in the document]
Comment (RodneyRichardson) [reply cyclonedx/model/__init__.py]: Ignore. Naming is consistent with existing __eq__ operator.

**Comment by RodneyRichardson** [comment]
> My take on the topic would have implemented `__lt__()` in each data model (like [so](https://github.com/CycloneDX/cyclonedx-javascript-library/blob/1.0-dev/src/models/component.ts#L59)), and used `sorted()` when it came to iterating in the normalizer/serializer (like [so](https://github.com/CycloneDX/cyclonedx-javascript-library/blob/1.0-dev/src/serialize/JSON/normalize.ts#L147-L149)).

That was my initial implementation, and it was OK for the XML serialization, but the JSON serialization doesn't explicitly iterate the collections when serializing so it didn't work. This approach is more consistent across serializers.

**Comment by RodneyRichardson** [comment]
@jkowalleck I have resolved all issues, and and CI jobs are passing now. This is ready for review/merge.

FYI: To solve the type hint issues in python < 3.8, I needed to put `SortedSet[T]` in quotes.

**Comment by jkowalleck** [comment]
Thanks for the feature, @RodneyRichardson 

will review the PR when i have time.
Also i want to see what @madpah has to say.

This means, that the PR will take it's time.

**Comment by jkowalleck** [comment]
bumped mypy and several other dependencies,
in the hope thhis might fix possible StaticCodeAnalysis false-positives.

could you rebase onto latest "main" branch, @RodneyRichardson ?

**Comment by jkowalleck** [comment]
> > could you rebase onto latest "main" branch, @RodneyRichardson ?
> 
> I have merged main into the PR/branch (just clicking the button in Github). Is this sufficient? Or do you need me to rebase the entire branch into a single commit? (I've not rebased on Github before).

no, thats fine.
The PR will be quashed by us on approval eventually.

**Review by madpah — APPROVED** [review]

**Comment by madpah** [comment]
@RodneyRichardson - many thanks for your great contribution here. All looks good to me, so we'll get this merged imminently!

**Pull request merged**